components/concept/engine/src/main/java/mozilla/components/concept/engine/profiler/ProfilerRuntime.kt

```diff
@@ -0,0 +1,151 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.profiler
+
+import java.lang.UnsupportedOperationException
+
+/**
+ * ProfilerRuntime is used to manage Firefox Profiler related features.
+ *
+ * If you want to add a profiler marker to mark a point in time (without a duration)
+ * you can directly use `engine.addProfilerMarker("marker name")`.
+ * Or if you want to provide more information, you can use
+ * `engine.addProfilerMarker("marker name", "extra information")`.
+ *
+ * If you want to add a profiler marker with a duration (with start and end time)
+ * you can use it like this, it will automatically get the end time inside the addProfilerMarker:
+ * ```
+ *     val startTime = engine.getProfilerTime()
+ *     ...some code you want to measure...
+ *     engine.addProfilerMarker("name", startTime)
+ * ```
+
+ * Or you can capture start and end time in somewhere, then add the marker in somewhere else:
+ * ```
+ *     val startTime = engine.getProfilerTime()
+ *     ...some code you want to measure (or end time can be collected in a callback)...
+ *     val endTime = engine.getProfilerTime()
+ *
+ *     ...somewhere else in the codebase...
+ *     engine.addProfilerMarker("name", startTime, endTime)
+ * ```
+ *
+ * Here's an [ProfilerRuntime.addProfilerMarker] example with all the possible parameters:
+ * ```
+ *     val startTime = engine.getProfilerTime()
+ *     ...some code you want to measure...
+ *     val endTime = engine.getProfilerTime()
+ *
+ *     ...somewhere else in the codebase...
+ *     engine.addProfilerMarker("name", startTime, endTime, "extra information")
+ * ```
+ *
+ * [ProfilerRuntime.isProfilerActive] method is handy when you want to get more information to
+ * add inside the marker, but you think it's going to be computationally heavy (and useless)
+ * when profiler is not running:
+ * ```
+ *     val startTime = engine.getProfilerTime()
+ *     ...some code you want to measure...
+ *     if (engine.isProfilerActive()) {
+ *         val info = aFunctionYouDoNotWantToCallWhenProfilerIsNotActive()
+ *         engine.addProfilerMarker("name", startTime, info)
+ *     }
+ * ```
+ */
+interface ProfilerRuntime {
+    /**
+     * Returns true if profiler is active and it's allowed the add markers.
+     * It's useful when it's computationally heavy to get startTime or the
+     * additional text for the marker. That code can be wrapped with
+     * isProfilerActive if check to reduce the overhead of it.
+     *
+     * @return true if profiler is active and safe to add a new marker.
+     */
+    fun isProfilerActive(): Boolean {
+        throw UnsupportedOperationException("Firefox Profiler support is not available in this engine")
```

Comment:
Currently I'm throwing some exceptions here. But maybe leaving these functions as no-op is a better option. Since these will be no-op if profiler is not running anyway. This way we can start adding profiler markers to the codebase without worrying about exceptions? What do you think?

Reply:
Do we need a default implementation at all? I'd expect `Engine` implementations that do not support `ProfilerRuntime` to just not declare it?

Is your intention here to make all Gecko variants of `Engine` implement them so that you can start using them in Fenix regardless what Gecko version is used? In that case I think I'd still recommend not having a default implementation here and no-op in the specific `GeckoEngine` that doesn't support it.